File: backend/src/web/WebFilterEngine.ts
import { resolveClassifierOutcome, resolveWebRuleSource } from './policy.js';
import type {
  CanonicalDomain,
  ClassifierResult,
  OpaqueFamilyId,
  VpnMetadataDecision,
  WebDecision,
  WebRuleSource,
} from './types.js';
import type { WebRuleRepository } from './WebRuleStore.js';

const REASON_CODES: Record<string, string> = {
  SECURITY_DENYLIST: 'blocked by a security threat rule',
  PARENT_ALLOWLIST: 'allowed by your family\'s allow list',
  PARENT_DENYLIST: 'blocked by your family\'s block list',
  CATEGORY_RULE: 'blocked by your family\'s content category rule',
  SCHEDULE_RULE: 'blocked by your family\'s schedule rule',
  CLASSIFIER: 'blocked by your family\'s explicit-content rule',
  DEFAULT: 'no matching rule; allowed by default',
  VPN_UNAVAILABLE: 'network filtering capability was unavailable for this request',
};

/**
 * doc 14's deterministic decision pipeline (steps B through I), implemented
 * as pure orchestration over injected ports -- this class holds no state of
 * its own. `classifierResult` is optional and, per doc 14 step F, only
 * meaningful when the domain was rendered in the PCA Safe Browser; callers
 * outside that context must omit it rather than pass a stale/irrelevant
 * result. `vpnDecision` is optional and reflects doc 14 layer 2 -- when
 * present with outcome UNAVAILABLE, this engine still evaluates rules but
 * never claims network-layer coverage it did not have (see coverage field).
 */
export class WebFilterEngine {
  private readonly rules: WebRuleRepository;

  constructor(rules: WebRuleRepository) {
    this.rules = rules;
  }

  async decide(
    familyId: OpaqueFamilyId,
    domain: CanonicalDomain,
    options: { classifierResult?: ClassifierResult; vpnDecision?: VpnMetadataDecision } = {},
  ): Promise<WebDecision> {
    const matched = await this.rules.findMatching(familyId, domain);
    const winner = resolveWebRuleSource(matched);
    if (winner !== null) {
      return {
        domain,
        outcome: winner.listType === 'ALLOW' ? 'ALLOW' : 'BLOCK',
        source: winner.source,
        reasonCode: REASON_CODES[winner.source as WebRuleSource],
        coverage: 'DOMAIN_ONLY',
      };
    }

    if (options.classifierResult) {
      const outcome = resolveClassifierOutcome(options.classifierResult);
      if (outcome !== 'ALLOW') {
        return {
          domain,
          outcome,
          source: 'CLASSIFIER',
          reasonCode: REASON_CODES.CLASSIFIER,
          coverage: 'FULL_URL',
        };
      }
    }

    if (options.vpnDecision && options.vpnDecision.domain === domain) {
      if (options.vpnDecision.outcome === 'BLOCKED') {
        return {
          domain,
          outcome: 'BLOCK',
          source: 'DEFAULT',
          reasonCode: REASON_CODES.PARENT_DENYLIST,
          coverage: options.vpnDecision.coverage === 'DESTINATION_ONLY' ? 'DESTINATION_ONLY' : 'DOMAIN_ONLY',
        };
      }
      if (options.vpnDecision.outcome === 'UNAVAILABLE') {
        return {
          domain,
          outcome: 'ALLOW',
          source: 'DEFAULT',
          reasonCode: REASON_CODES.VPN_UNAVAILABLE,
          coverage: 'DOMAIN_ONLY',
        };
      }
    }

    return {
      domain,
      outcome: 'ALLOW',
      source: 'DEFAULT',
      reasonCode: REASON_CODES.DEFAULT,
      coverage: 'DOMAIN_ONLY',
    };
  }
}

File: android/app/src/main/java/org/pca/app/feature/webprotection/engine/WebFilterEngine.kt
package org.pca.app.feature.webprotection.engine

import org.pca.app.feature.webprotection.policy.CanonicalDomain
import org.pca.app.feature.webprotection.policy.ClassifierResult
import org.pca.app.feature.webprotection.policy.OpaqueFamilyId
import org.pca.app.feature.webprotection.policy.VpnDecisionCoverage
import org.pca.app.feature.webprotection.policy.VpnDecisionOutcome
import org.pca.app.feature.webprotection.policy.VpnMetadataDecision
import org.pca.app.feature.webprotection.policy.WebDecision
import org.pca.app.feature.webprotection.policy.WebDecisionCoverage
import org.pca.app.feature.webprotection.policy.WebDecisionOutcome
import org.pca.app.feature.webprotection.policy.WebDecisionSource
import org.pca.app.feature.webprotection.policy.WebLocale
import org.pca.app.feature.webprotection.policy.WebReasonId
import org.pca.app.feature.webprotection.policy.resolveClassifierOutcome
import org.pca.app.feature.webprotection.policy.resolveWebRuleSource
import org.pca.app.feature.webprotection.policy.toDecisionSource
import org.pca.app.feature.webprotection.policy.translateWebReason

/**
 * doc 14's deterministic decision pipeline (steps B through I), mirrored
 * line-for-line from `backend/src/web/WebFilterEngine.ts`: this class holds
 * no state of its own, is pure orchestration over an injected
 * [WebRuleRepository], and runs entirely on-device -- satisfying doc 14's
 * offline requirement directly (there is no network call anywhere in this
 * file, by construction, not just by policy).
 *
 * `classifierResult` is optional and, per doc 14 step F, only meaningful
 * when the domain was rendered in the PCA Safe Browser; callers outside that
 * context must omit it. `vpnDecision` is optional and reflects doc 14 layer 2
 * -- when present with outcome UNAVAILABLE, this engine still evaluates
 * rules but never claims network-layer coverage it did not have (the
 * resulting [WebDecision.coverage] stays DOMAIN_ONLY, never FULL_URL).
 *
 * The "remote classifier unavailable must never silently degrade to allow"
 * requirement is satisfied structurally: [decide] never treats a missing/
 * omitted [ClassifierResult] as an ALLOW signal on its own -- it is simply
 * skipped, and evaluation continues to the VPN-metadata and deterministic
 * default-allow steps exactly as if no classifier existed at all. There is
 * no code path in this class that maps "classifier unavailable" to a BLOCK
 * bypass.
 */
class WebFilterEngine(private val rules: WebRuleRepository) {

    fun decide(
        familyId: OpaqueFamilyId,
        domain: CanonicalDomain,
        classifierResult: ClassifierResult? = null,
        vpnDecision: VpnMetadataDecision? = null,
        locale: WebLocale = WebLocale.EN,
    ): WebDecision {
        val matched = rules.findMatching(familyId, domain)
        val winner = resolveWebRuleSource(matched)
        if (winner != null) {
            val outcome = if (winner.listType.name == "ALLOW") WebDecisionOutcome.ALLOW else WebDecisionOutcome.BLOCK
            val source = winner.source.toDecisionSource()
            return build(domain, outcome, source, source.toReasonId(), WebDecisionCoverage.DOMAIN_ONLY, locale)
        }

        if (classifierResult != null) {
            val outcome = resolveClassifierOutcome(classifierResult)
            if (outcome != WebDecisionOutcome.ALLOW) {
                return build(domain, outcome, WebDecisionSource.CLASSIFIER, WebReasonId.CLASSIFIER, WebDecisionCoverage.FULL_URL, locale)
            }
        }

        if (vpnDecision != null && vpnDecision.domain == domain) {
            if (vpnDecision.outcome == VpnDecisionOutcome.BLOCKED) {
                val coverage = if (vpnDecision.coverage == VpnDecisionCoverage.DESTINATION_ONLY) {
                    WebDecisionCoverage.DESTINATION_ONLY
                } else {
                    WebDecisionCoverage.DOMAIN_ONLY
                }
                return build(domain, WebDecisionOutcome.BLOCK, WebDecisionSource.DEFAULT, WebReasonId.PARENT_DENYLIST, coverage, locale)
            }
            if (vpnDecision.outcome == VpnDecisionOutcome.UNAVAILABLE) {
                return build(domain, WebDecisionOutcome.ALLOW, WebDecisionSource.DEFAULT, WebReasonId.VPN_UNAVAILABLE, WebDecisionCoverage.DOMAIN_ONLY, locale)
            }
        }

        return build(domain, WebDecisionOutcome.ALLOW, WebDecisionSource.DEFAULT, WebReasonId.DEFAULT, WebDecisionCoverage.DOMAIN_ONLY, locale)
    }

    private fun WebDecisionSource.toReasonId(): WebReasonId = when (this) {
        WebDecisionSource.SECURITY_DENYLIST -> WebReasonId.SECURITY_DENYLIST
        WebDecisionSource.PARENT_ALLOWLIST -> WebReasonId.PARENT_ALLOWLIST
        WebDecisionSource.PARENT_DENYLIST -> WebReasonId.PARENT_DENYLIST
        WebDecisionSource.CATEGORY_RULE -> WebReasonId.CATEGORY_RULE
        WebDecisionSource.SCHEDULE_RULE -> WebReasonId.SCHEDULE_RULE
        WebDecisionSource.CLASSIFIER -> WebReasonId.CLASSIFIER
        WebDecisionSource.DEFAULT -> WebReasonId.DEFAULT
    }

    private fun build(
        domain: CanonicalDomain,
        outcome: WebDecisionOutcome,
        source: WebDecisionSource,
        reasonId: WebReasonId,
        coverage: WebDecisionCoverage,
        locale: WebLocale,
    ): WebDecision = WebDecision(
        domain = domain,
        outcome = outcome,
        source = source,
        reasonId = reasonId,
        reasonCode = translateWebReason(reasonId, locale),
        coverage = coverage,
    )
}
